### DISCUSSION TITLE
Narrow typing the QueryKey type definition

### DISCUSSION THREAD

**Discussion by dominictwlee** [TanStack/query#1849] [Ideas] [answered]
I was wondering whether having the string key as the first element of an array key is a personal preference, or something that should be enforced as a convention?

It seems all the examples in the docs enforce the idea that the string key should be used:
```
useQuery(['todos', status, page], ...)
useQuery(['todos', page, status], ...)
useQuery(['todos', undefined, page, status], ...)
```

If that's the case, would it be a good idea to narrow type the definition to something like:
```
export type QueryKey = string | [string, ...unknown[]]
```

**Comment by TkDodo** [comment, marked as answer]
keys can be string or any array that people want, but the type inference would need some love. `unknown[]` is actually way too broad :)
